### PR TITLE
[`Mask2Former`] Move normalization for numerical stability

### PR DESCRIPTION
**Pull request by amyeroberts**
# What does this PR do?

Moving the normalization before the matmul operation makes the calculation more stable and less likely to overflow. 

Same differences introduce in #26086 which was closed after becoming stale.

### CODE REVIEW

File: src/transformers/models/mask2former/modeling_mask2former.py [hunk 1]
```diff
@@ -373,13 +373,13 @@ def pair_wise_sigmoid_cross_entropy_loss(inputs: torch.Tensor, labels: torch.Ten
...
-    cross_entropy_loss_pos = criterion(inputs, torch.ones_like(inputs))
-    cross_entropy_loss_neg = criterion(inputs, torch.zeros_like(inputs))
+    cross_entropy_loss_pos = criterion(inputs, torch.ones_like(inputs)) / height_and_width
+    cross_entropy_loss_neg = criterion(inputs, torch.zeros_like(inputs)) / height_and_width
 
     loss_pos = torch.matmul(cross_entropy_loss_pos, labels.T)
     loss_neg = torch.matmul(cross_entropy_loss_neg, (1 - labels).T)
     loss = loss_pos + loss_neg
-    loss = loss / height_and_width
+    loss = loss
```
Comment (amyeroberts) [inline src/transformers/models/mask2former/modeling_mask2former.py:382]:
```suggestion
```

File: src/transformers/models/oneformer/modeling_oneformer.py [hunk 2]
```diff
@@ -198,13 +198,13 @@ def pair_wise_sigmoid_cross_entropy_loss(inputs: torch.Tensor, labels: torch.Ten
...
-    cross_entropy_loss_pos = criterion(inputs, torch.ones_like(inputs))
-    cross_entropy_loss_neg = criterion(inputs, torch.zeros_like(inputs))
+    cross_entropy_loss_pos = criterion(inputs, torch.ones_like(inputs)) / height_and_width
+    cross_entropy_loss_neg = criterion(inputs, torch.zeros_like(inputs)) / height_and_width
 
     loss_pos = torch.matmul(cross_entropy_loss_pos, labels.T)
     loss_neg = torch.matmul(cross_entropy_loss_neg, (1 - labels).T)
     loss = loss_pos + loss_neg
-    loss = loss / height_and_width
+    loss = loss
```
Comment (amyeroberts) [inline src/transformers/models/oneformer/modeling_oneformer.py:207]:
```suggestion
```

File: src/transformers/models/mask2former/modeling_mask2former.py [hunk 3]
```diff
@@ -373,13 +373,12 @@ def pair_wise_sigmoid_cross_entropy_loss(inputs: torch.Tensor, labels: torch.Ten
...
 
     criterion = nn.BCEWithLogitsLoss(reduction="none")
-    cross_entropy_loss_pos = criterion(inputs, torch.ones_like(inputs))
-    cross_entropy_loss_neg = criterion(inputs, torch.zeros_like(inputs))
+    cross_entropy_loss_pos = criterion(inputs, torch.ones_like(inputs)) / height_and_width
+    cross_entropy_loss_neg = criterion(inputs, torch.zeros_like(inputs)) / height_and_width
 
     loss_pos = torch.matmul(cross_entropy_loss_pos, labels.T)
     loss_neg = torch.matmul(cross_entropy_loss_neg, (1 - labels).T)
     loss = loss_pos + loss_neg
```
Comment (ydshieh) [inline src/transformers/models/mask2former/modeling_mask2former.py:381]: fine for me. But the name might be a bit misleading (I am not 100% sure): does `criterion(inputs, torch.ones_like(inputs)) / height_and_width` really representing `cross_entropy_loss`?

I am not checking the full context here, but personally, I might just do something like

```
    cross_entropy_loss_pos = criterion(inputs, torch.ones_like(inputs)) 
    cross_entropy_loss_neg = criterion(inputs, torch.zeros_like(inputs)) 

    loss_pos = torch.matmul(cross_entropy_loss_pos / height_and_width, labels.T)
    loss_neg = torch.matmul(cross_entropy_loss_neg / height_and_width, (1 - labels).T)
```
to avoid (if any) possible confusion.

**Review by ydshieh — APPROVED** [review]
Thanks!
